File: src/leetCode/Ex50.java
package leetCode;

public class Ex50 {
    public double myPow(double x, int n) {
        if (n == 1) return x;
        if (n == 0) return 1;
        if(n == -2147483648){
            if (x!=1.0&&x!=-1.0) return 0.0;
            else return 1.0;
        }
        if (n < 0) return 1 / myPow(x, -n);

        int[] base = new int[32];
        int t = n;
        int i = 0;
        while (t > 0 && i < 32) {
            base[i++] = t & 1;
            t = t >>> 1;
        }
        double temp = x;
        double res = 1.0;
        for (int j = 0; j < 32; j++) {
            if (base[j] == 1) {
                res *= temp;
            }
            temp = temp * temp;
        }

        return res;
    }

    public static void main(String[] args) {
        System.out.println(new Ex50().myPow(2, 10));
    }
}

File: src/leetCode/ex50.kt
package leetCode

fun myPow(x: Double, n: Int): Double {
    if (n == 0) {
        return 1.0
    }
    var times = if (n > 0) n.toLong() else -n.toLong()
    val xx = if (n > 0) x else 1 / x
    var res = xx
    while (times > 1) {
        res = if (times % 2 == 1.toLong()) res * res * xx
        else res * res
        times /= 2
    }
    return res
}

fun main(args: Array<String>) {
    intArrayOf( -5).forEach { myPow(8.84372, it).apply(::println) }
}